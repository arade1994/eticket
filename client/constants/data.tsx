import {
  FaBell,
  FaChartLine,
  FaCog,
  FaComments,
  FaMapMarkerAlt,
  FaTags,
  FaTicketAlt,
  FaUsers,
} from "react-icons/fa";

export const HOMEPAGE_FEATURES = [
  {
    icon: <FaTicketAlt />,
    title: "Easy Ticket Management",
    description:
      "Buy and sell tickets seamlessly with our secure payment system and real-time updates.",
    color: "#006064",
  },
  {
    icon: <FaUsers />,
    title: "Social Features",
    description:
      "Connect with other event enthusiasts, share experiences, and track each other's events.",
    color: "#2E7D32",
  },
  {
    icon: <FaMapMarkerAlt />,
    title: "Event Details",
    description:
      "Get comprehensive event information including location, timing, and venue details.",
    color: "#C2185B",
  },
  {
    icon: <FaChartLine />,
    description:
      "Track event popularity and get insights with our advanced analytics dashboard.",
    color: "#1976D2",
  },
  {
    icon: <FaBell />,
    title: "Smart Notifications",
    description:
      "Stay updated with personalized notifications for events, messages, and updates.",
    color: "#F57C00",
  },
  {
    icon: <FaComments />,
    title: "Chat System",
    description:
      "Communicate directly with other users about events and tickets.",
    color: "#7B1FA2",
  },
  {
    icon: <FaCog />,
    title: "Customizable Settings",
    description:
      "Personalize your experience with custom filters and display preferences.",
    color: "#455A64",
  },
  {
    icon: <FaTags />,
    title: "Categories & Tags",
    description:
      "Organize events with our comprehensive category and subcategory system.",
    color: "#D32F2F",
  },
];
