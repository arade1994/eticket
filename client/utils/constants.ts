export const TICKET_CATEGORIES = [
  "Sport game",
  "Music event",
  "Theatre play",
  "Public transport",
  "Online tickets",
];
