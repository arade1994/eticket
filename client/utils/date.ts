/**
 * Calculates expiration date (which is 15 minutes long) starting from the current moment
 * @returns Expiration date
 */
export const calculateExpirationDate = () => {
  const expiration = new Date();
  return expiration.setSeconds(expiration.getSeconds() + 15 * 60);
};

/**
 * Formats date string to a hh:mm format
 * @param date - string date
 * @returns date in hh:mm format
 */
export const formatDateToHHMM = (date: string) => {
  const dateObject = new Date(date);

  return `${dateObject.getHours()}:${dateObject.getMinutes()}`;
};
