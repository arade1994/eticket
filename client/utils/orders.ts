import { type Order } from "../types/order";

/**
 * Filters orders by ticket title, status and user that created the order
 * @param orders - list of orders
 * @param searchText - ticket title
 * @param userId - id of user that created the order
 * @param status - selected order status
 * @returns filtered list of orders
 */
export const filterOrders = (
  orders: Order[],
  searchText: string,
  userId: string,
  status: string
) => {
  let filteredOrders = [...orders];

  if (searchText.trim() !== "")
    filteredOrders = filteredOrders.filter((order) =>
      order.ticket.title.toLowerCase().includes(searchText.toLowerCase())
    );

  if (userId !== "-")
    filteredOrders = filteredOrders.filter((order) => order.userId === userId);

  if (status !== "-")
    filteredOrders = filteredOrders.filter((order) => order.status === status);

  return filteredOrders;
};
