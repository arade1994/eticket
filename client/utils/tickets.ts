import { Ticket } from "../types/ticket";

export const filterTickets = (
  tickets: Ticket[],
  searchText: string,
  category: string,
  userId: string
) => {
  let filteredTickets = [...tickets];

  if (searchText.trim() !== "")
    filteredTickets = filteredTickets.filter((ticket) =>
      ticket.title.toLowerCase().includes(searchText.toLowerCase())
    );

  if (category !== "-")
    filteredTickets = filteredTickets.filter(
      (ticket) => ticket.category === category
    );

  if (userId !== "-")
    filteredTickets = filteredTickets.filter(
      (ticket) => ticket.userId === userId
    );

  return filteredTickets;
};
