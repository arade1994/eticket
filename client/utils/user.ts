import { Ticket } from "../types/ticket";
import { User } from "../types/user";

export const getNumOfCreatedTickets = (user: User, tickets: Ticket[]) =>
  tickets?.filter((ticket) => ticket.userId === user.id)?.length;
