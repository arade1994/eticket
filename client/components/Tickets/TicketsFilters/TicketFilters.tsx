import React, { type ChangeEvent, useMemo } from "react";
import Select, { type SingleValue } from "react-select";

import { type User } from "../../../types/user";
import { ticketCategoriesOptions } from "../../../utils/constants";

import classes from "./TicketFilters.module.scss";

interface Props {
  category: {
    value: string;
    label: string;
  };
  searchText: string;
  selectedUser: {
    value: string;
    label: string;
  };
  users: User[];
  onChangeCategory: (
    option: SingleValue<{
      value: string;
      label: string;
    }>
  ) => void;
  onChangeSearchText: (event: ChangeEvent<HTMLInputElement>) => void;
  onResetFilters: () => void;
  onSelectUser: (
    option: SingleValue<{
      value: string;
      label: string;
    }>
  ) => void;
}

const TicketFilters = ({
  category,
  searchText,
  selectedUser,
  users,
  onChangeCategory,
  onChangeSearchText,
  onResetFilters,
  onSelectUser,
}: Props) => {
  const userOptions = useMemo(
    () =>
      users.map((user) => ({
        value: user.id,
        label: `${user.firstName} ${user.lastName}`,
      })),
    [users]
  );

  const isFiltersApplied = useMemo(
    () =>
      searchText.trim() !== "" ||
      category.value !== "-" ||
      selectedUser.value !== "-",
    [searchText, category.value, selectedUser.value]
  );

  return (
    <div className={classes.ticketFilters}>
      <input
        id="title"
        placeholder="Insert title"
        type="text"
        value={searchText}
        onChange={onChangeSearchText}
      />
      <Select
        id="categorySelect"
        options={[
          { value: "-", label: "Any category" },
          ...ticketCategoriesOptions,
        ]}
        value={category}
        onChange={onChangeCategory}
      />
      <Select
        id="userSelect"
        options={[{ value: "-", label: "Any user" }, ...userOptions]}
        value={selectedUser}
        onChange={onSelectUser}
      />
      {isFiltersApplied && (
        <div className={classes.resetFilters} onClick={onResetFilters}>
          Reset filters
        </div>
      )}
    </div>
  );
};

export default TicketFilters;
