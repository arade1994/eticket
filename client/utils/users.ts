import { type Ticket } from "../types/ticket";
import { type Rating, type User } from "../types/user";

/**
 * Returns number of created tickets per user
 * @param userId - specific user
 * @param tickets - list of tickets
 * @returns number of tickets
 */
export const getNumOfCreatedTickets = (userId: string, tickets: Ticket[]) =>
  tickets.filter((ticket) => ticket.userId === userId)?.length;

/**
 * Gets all the ratings for one specific user
 * @param userId - specific user
 * @param ratings - ratings list
 * @returns - all ratings for user
 */
export const getUserRatings = (userId: string, ratings: Rating[]) => {
  return ratings.filter((rating) => rating.ratedUserId === userId);
};

/**
 * Calculates user final rating number
 * @param userId - specific user
 * @param ratings - ratings list
 * @returns - final rating
 */
export const getUserRating = (userId: string, ratings: Rating[]) => {
  const userRates = getUserRatings(userId, ratings)?.map(
    (rating) => rating.rate
  );
  const ratesSum = userRates.reduce((a, b) => a + b, 0);
  if (!ratesSum || isNaN(ratesSum)) return "-";
  return (ratesSum / userRates.length).toFixed(1);
};

/**
 * Returns an information whether user is rated or not by the currently logged in user
 * @param userId - specific user
 * @param ratings - list of ratings
 * @param currentUserId - logged in user
 * @returns - boolean value for rated status of a user
 */
export const getIsUserRated = (
  userId: string,
  ratings: Rating[],
  currentUserId: string
) =>
  !!ratings.filter(
    (rating) => rating.ratedUserId === userId && rating.userId === currentUserId
  )?.length;

/**
 * Filters users list by first and last name
 * @param users - list of users
 * @param searchText - first name or last name text
 * @returns - list of filtered users
 */
export const filterUsers = (users: User[], searchText: string) => {
  let filteredUsers = [...users];

  if (searchText.trim() !== "")
    filteredUsers = filteredUsers.filter(
      (user) =>
        user.firstName.toLowerCase().includes(searchText) ||
        user.lastName.toLowerCase().includes(searchText)
    );

  return filteredUsers;
};
