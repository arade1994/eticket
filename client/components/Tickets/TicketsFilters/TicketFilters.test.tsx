import { render, screen, waitFor, within } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import TicketFilters from "./TicketFilters";
import { ticketCategoriesOptions } from "../../../utils/constants";
import mockDb from "../../../mock/db.json";
import userEvent from "@testing-library/user-event";

const mockUsers = mockDb.users;
const mockSelectedUser = {
  value: "-",
  label: "Any user",
};

const mockOnChangeCategory = vi.fn();
const mockOnChangeSearchText = vi.fn();
const mockOnResetFilters = vi.fn();
const mockOnSelectUser = vi.fn();

const renderTicketFilters = (searchText: string) =>
  render(
    <TicketFilters
      category={ticketCategoriesOptions[0]}
      searchText={searchText}
      selectedUser={mockSelectedUser}
      users={mockUsers}
      onChangeCategory={mockOnChangeCategory}
      onChangeSearchText={mockOnChangeSearchText}
      onResetFilters={mockOnResetFilters}
      onSelectUser={mockOnSelectUser}
    />
  );

describe("<TicketFilters />", () => {
  test("it should match a snapshot when rendered by default", () => {
    const { baseElement } = renderTicketFilters("");

    expect(baseElement).toMatchSnapshot();
  });

  test("it should render title filter", async () => {
    renderTicketFilters("");

    const titleInput = screen.getByTestId("title") as HTMLInputElement;
    expect(titleInput).toBeDefined();
    await userEvent.type(titleInput, "Test");
    expect(mockOnChangeSearchText).toHaveBeenCalled();
  });

  test("it should render category select", async () => {
    renderTicketFilters("");

    const categorySelect = within(
      screen.getByTestId("categorySelect")
    ).getByText("Sport game");
    expect(categorySelect).toBeDefined();
    expect(categorySelect.textContent).toEqual("Sport game");
    await userEvent.click(categorySelect);
    await waitFor(() => screen.getByText("Theatre play"));
    await userEvent.click(screen.getByText("Theatre play"));
    expect(mockOnChangeCategory).toHaveBeenCalled();
  });

  test("it should render user select", async () => {
    renderTicketFilters("");

    const userSelect = within(screen.getByTestId("userSelect")).getByText(
      "Any user"
    );
    expect(userSelect).toBeDefined();
    expect(userSelect.textContent).toEqual("Any user");
    await userEvent.click(userSelect);
    await waitFor(() => screen.getByText("Ante Rade"));
    await userEvent.click(screen.getByText("Ante Rade"));
    expect(mockOnSelectUser).toHaveBeenCalled();
  });

  test("it should render reset filters button when filters are applied", async () => {
    renderTicketFilters("Test");

    const resetFiltersBtn = screen.getByText("Reset filters");
    expect(resetFiltersBtn).toBeDefined();
    await userEvent.click(resetFiltersBtn);
    expect(mockOnResetFilters).toHaveBeenCalled();
  });
});
