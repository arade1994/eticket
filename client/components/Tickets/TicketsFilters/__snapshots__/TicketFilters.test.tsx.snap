// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TicketFilters /> > it should match a snapshot when rendered by default 1`] = `
<body>
  <div>
    <div
      class="_ticketFilters_f019c6"
    >
      <input
        id="title"
        placeholder="Insert title"
        type="text"
        value=""
      />
      <div
        class="css-b62m3t-container"
        id="categorySelect"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="css-13cymwt-control"
        >
          <div
            class="css-hlgwow"
          >
            <div
              class="css-1dimb5e-singleValue"
            >
              Sport game
            </div>
            <div
              class="css-19bb58m"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="css-1wy0on6"
          >
            <span
              class="css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-b62m3t-container"
        id="userSelect"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="css-13cymwt-control"
        >
          <div
            class="css-hlgwow"
          >
            <div
              class="css-1dimb5e-singleValue"
            >
              Any user
            </div>
            <div
              class="css-19bb58m"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-3-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="css-1wy0on6"
          >
            <span
              class="css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="_itemsPerPageSelect_f019c6 css-b62m3t-container"
        id="itemsPerPageSelect"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-4-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="css-13cymwt-control"
        >
          <div
            class="css-hlgwow"
          >
            <div
              class="css-1dimb5e-singleValue"
            >
              8 per page
            </div>
            <div
              class="css-19bb58m"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-4-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="css-1wy0on6"
          >
            <span
              class="css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <button
        class="_resetFilters_f019c6"
      >
        <svg
          class="_resetIcon_f019c6"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 352 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          />
        </svg>
        Reset filters
      </button>
    </div>
  </div>
</body>
`;
